tests/visualization_tests/matplotlib_tests/test_pareto_front.py

```diff
@@ -164,36 +191,59 @@ def test_plot_pareto_front_3d(
     study.enqueue_trial({"x": 1, "y": 1, "z": 1})
-    study.enqueue_trial({"x": 1, "y": 0, "z": 1})
+    study.enqueue_trial({"x": 1, "y": 0, "z": 2})
     study.enqueue_trial({"x": 1, "y": 1, "z": 0})
     study.optimize(
-        lambda t: [t.suggest_int("x", 0, 1), t.suggest_int("y", 0, 1), t.suggest_int("z", 0, 1)],
-        n_trials=3,
+        lambda t: [t.suggest_int("x", 0, 1), t.suggest_int("y", 0, 2), t.suggest_int("z", 0, 2)],
+        n_trials=4,
     )
 
+    constraints_func: Optional[Callable[[FrozenTrial], Sequence[float]]]
+    if use_constraints_func:
+        # (x, y, z) = (1, 1, 0) is infeasible; others are feasible.
+        def constraints_func(t: FrozenTrial) -> Sequence[float]:
+            return (
+                [1.0]
+                if t.params["x"] == 1 and t.params["y"] == 1 and t.params["z"] == 0
+                else [-1.0]
+            )
+
+    else:
+        constraints_func = None
+
     figure = plot_pareto_front(
         study=study,
         include_dominated_trials=include_dominated_trials,
         axis_order=axis_order,
+        constraints_func=constraints_func,
         targets=targets,
     )
+    actual_axis_order = axis_order or [0, 1, 2]
     assert len(figure.get_lines()) == 0
 
-    if axis_order is not None:
-        pareto_front_points = np.array([[1.0, 0.0, 1.0], [1.0, 1.0, 0.0]])[:, axis_order][:, 0:2]
+    if use_constraints_func:
+        if include_dominated_trials:
+            # The enqueue order of trial is: infeasible, feasible non-best, then feasible best.
+            assert len(figure.collections) == 3
+            data = [(1, 1, 1, 1), (1, 0, 1, 1), (1, 2, 2, 0)]  # type: ignore
+        else:
+            # The enqueue order of trial is: infeasible, feasible.
+            assert len(figure.collections) == 2
+            data = [(1, 1, 1), (1, 0, 1), (1, 2, 0)]  # type: ignore
     else:
-        pareto_front_points = np.array([[1.0, 0.0, 1.0], [1.0, 1.0, 0.0]])[:, 0:2]
-    assert pareto_front_points.shape == (2, 2)
-
-    path_offsets = list(map(lambda pc: pc.get_offsets(), figure.findobj(PathCollection)))
-    exists_pareto_front = any(
-        map(lambda po: allclose_as_set(po, pareto_front_points), path_offsets)
-    )
-    exists_dominated_trials = any(map(lambda po: allclose_as_set(po, [[1.0, 1.0]]), path_offsets))
-    assert exists_pareto_front
-    if include_dominated_trials:
-        assert exists_dominated_trials
+        if include_dominated_trials:
+            # The last elements come from dominated trial that is enqueued firstly.
+            assert len(figure.collections) == 2
+            data = [(1, 1, 1, 1), (0, 1, 1, 1), (2, 0, 2, 1)]  # type: ignore
+        else:
+            assert len(figure.collections) == 1
+            data = [(1, 1), (0, 1), (2, 0)]  # type: ignore
+
+    _check_data(figure, "x", data[actual_axis_order[0]])
+    _check_data(figure, "y", data[actual_axis_order[1]])
+    # _check_data(figure, "z", data[actual_axis_order[2]])
```

Comment:
I cannot find how to fetch `z` data from figure.
It seems that z hasn't been tested before.

Reply:
Maybe this is a follow-up task. Could you add TODO comment about it?